miri-script/src/commands.rs

```diff
@@ -0,0 +1,652 @@
+        }
+
+        Ok(output?)
+    }
+    /// Returns a toolchain override for use in command line (`+{toolchain_name}`).
+    fn get_toolchain_override(&self) -> OsString {
+        format!("+{}", &self.active_toolchain).into()
+    }
+    fn build_package(
+        // Path to Cargo.toml file of a package to build.
+        path: &OsStr,
+        toolchain: &OsStr,
+        extra_flags: &[String],
+        args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    ) -> Result<()> {
+        let sh = Shell::new()?;
+        cmd!(sh, "cargo {toolchain} build {extra_flags...} --manifest-path {path} {args...}")
+            .run()?;
+        Ok(())
+    }
+    fn shell_with_miri_env(&self) -> Result<Shell> {
+        let sh = shell_with_parent_env()?;
+        for (k, v) in &self.env {
+            sh.set_var(k, v);
+        }
+
+        Ok(sh)
+    }
+}
+
+impl MiriRunner<'_> {
+    fn bench(&self, benches: &[OsString]) -> Result<()> {
+        // The hyperfine to use
+        let hyperfine = std::env::var("HYPERFINE");
+        let hyperfine = hyperfine.as_deref().unwrap_or("hyperfine -w 1 -m 5 --shell=none");
+        let hyperfine = shell_words::split(hyperfine).unwrap();
+        let Some((program_name, args)) = hyperfine.split_first() else {
+            bail!("Expected HYPERFINE environment variable to be non-empty");
+        };
+        // Make sure we have an up-to-date Miri installed
+        Self::exec_inner(&Subcommands::Install { flags: vec![] }, false)?;
+        let benches_dir = path!(self.miri_dir / "bench-cargo-miri");
+        let benches = if benches.is_empty() {
+            std::fs::read_dir(&benches_dir)?
+                .filter_map(|path| {
+                    path.ok()
+                        .filter(|dir| dir.file_type().map(|t| t.is_dir()).unwrap_or(false))
+                        .map(|p| p.file_name())
+                })
+                .collect()
+        } else {
+            benches.to_owned()
+        };
+        let sh = shell_with_parent_env()?;
+        let toolchain = &self.active_toolchain;
+        // Run the requested benchmarks
+        for bench in benches {
+            let current_bench_dir = path!(benches_dir / bench / "Cargo.toml");
+            cmd!(
+                sh,
+                "{program_name} {args...} 'cargo +'{toolchain}' miri run --manifest-path '{current_bench_dir}''"
+            )
+            .run()?;
+        }
+        Ok(())
+    }
+
+    fn toolchain(&self, flags: &[OsString]) -> Result<()> {
+        // Make sure rustup-toolchain-install-master is installed.
+        which::which("rustup-toolchain-install-master").context("Please install rustup-toolchain-install-master by running 'cargo install rustup-toolchain-install-master'")?;
+        let sh = shell_with_parent_env()?;
+        sh.change_dir(&self.miri_dir);
+        let new_commit =
+            Some(sh.read_file(path!(self.miri_dir / "rust-version"))?.trim().to_owned());
```

Comment:
```suggestion
            Some(sh.read_file("rust-version")?.trim().to_owned());
```
why use `sh.read_file` if you don't make use of the relative working dir...?

Reply:
Hm, this does make use of relative working dir according to [xshell's doc](https://docs.rs/xshell/latest/xshell/).
Specifically:
`[Shell::read_file](https://docs.rs/xshell/latest/xshell/struct.Shell.html#method.read_file) works like [std::fs::read_to_string](https://doc.rust-lang.org/nightly/std/fs/fn.read_to_string.html), but paths are relative to the current directory of the [Shell](https://docs.rs/xshell/latest/xshell/struct.Shell.html).`

Reply:
Yeah but you used `self.miri_dir / "rust-version"`, which is an absolute path, so the `sh` working dir doesn't matter. With my patch it is being used.

Reply:
Oh, you're right. 😅